### PR TITLE
Spark: Implement an adapter to wrap Row into DataFile

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR implements an adapter to wrap Spark `Row`s into `DataFile`s so that we can write Spark `Row`s to a manifest.

### CODE REVIEW

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  private final SparkStructLike wrappedPartition;
+  private Row wrapped;
+
+  public SparkDataFile(Types.StructType type) {
+    this.lowerBoundsType = type.fieldType("lower_bounds");
+    this.upperBoundsType = type.fieldType("upper_bounds");
+    this.keyMetadataType = type.fieldType("key_metadata");
+    this.wrappedPartition = new SparkStructLike(type.fieldType("partition").asStructType());
+    // the partition field is absent for unpartitioned tables
+    this.fieldShift = wrappedPartition.size() != 0 ? 1 : 0;
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:48]: As much as I don't like to use the index, referring by column name seems to incur an additional penalty as we will resolve the column name every time (based on what I see in Spark `Row`).

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  @Override
+  public long recordCount() {
+    return wrapped.getAs(fieldShift + 2);
+  }
+
+  @Override
+  public long fileSizeInBytes() {
+    return wrapped.getAs(fieldShift + 3);
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:82]: I am using `getAs` instead of `getLong` as `getLong` and other methods specific to primitives perform a check if the value is null. The downside might be boxing/unboxing.

**Comment by aokolnychyi** [comment]
@rdblue, do we need to update `TestSparkReadProjection` to use this conversion logic too? It seems that test won't handle dates and timestamps as Spark `Row` returns `java.sql.Date` and `java.sql.Timestamp` respectively.

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java (hunk 2 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:82]: Actually, Spark row uses `Array[Any]` which I assume allocates an array of `Object`s. So, Spark stores boxed versions and we have to unbox it anyway. If we used `getLong`, we would also perform a null check.

File: build.gradle [hunk 3]
```diff
@@ -307,6 +307,7 @@ project(':iceberg-spark') {
     compile project(':iceberg-api')
     compile project(':iceberg-common')
     compile project(':iceberg-core')
+    compile project(':iceberg-data')
```
Comment (rdblue) [inline build.gradle:310]: Why is this necessary? It looks like all of the code here is added to Spark.

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java (hunk 1 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:48]: If we assume that the type passed in here matches the Row, then we should be able to derive the ordinals from that type instead of hard-coding them. I think this would produce an int[] of expected offset to actual offset, like we use in the Avro generics. What do you think?

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:48]: Otherwise, I think it would probably be worth the performance penalty to ensure correctness.

File: spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Map;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.util.DateTimeUtils;
+
+public class SparkValueConverter {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java:37]: It would be nice to have some Javadoc about this, like that values are converted to Iceberg's internal representation.

File: build.gradle (hunk 3 above)
Comment (aokolnychyi) [reply build.gradle:310]: `SparkValueConverter` uses the following classes:

```
import org.apache.iceberg.data.GenericRecord;
import org.apache.iceberg.data.Record;
```

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:48]: @rdblue, do you mean passing the Spark type alongside the Iceberg type and building a mapping in the constructor?

File: spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java:37]: Yep, will add.

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java (hunk 1 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java:48]: Ideally, yes. But if that's not possible we know that the DataFile that will be passed in comes from a metadata table (because we control the implementations that use this) so it would be okay to assume that we know what the Spark schema is.

File: spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java:40]: Done.

File: spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java]: Okay, I've updated this to handle arbitrary projections.

**Comment by aokolnychyi** [comment]
@rdblue, what about updating `TestSparkReadProjection` as described [here](https://github.com/apache/incubator-iceberg/pull/937#issuecomment-615540803)?

**Comment by rdblue** [comment]
> what about updating TestSparkReadProjection

I think it's reasonable to update the test, but not required. It's always good to reduce duplicated code, but we don't actually need that to support date and timestamp because those aren't tested.

**Comment by rdblue** [comment]
I ran `./gradlew check` locally and everything looks good so I merged this.
